### ISSUE TITLE
Quitting on MacOS fails to trigger an AppExit event

### ISSUE THREAD

**Issue by roboticdemonowl** [bevyengine/bevy#9499]
## Bevy version

0.11.0

## Relevant system information

This is a MacOS bug.

## What you did

I quit the application with the keyboard shortcut "Command+q".

## What went wrong

I was expecting it to end the program gracefully, trigger an AppExit event, and call the implemented Drop functions for the components that I set up.

It actually just ends the application's process without any of this cleanup functionality having an opportunity to run. It works as expected when I close the primary window or trigger the AppExit event myself, but not when I call the application's Quit menu item.

## Additional information

Ideally, quitting the application through the application menu would simply fire the AppExit event instead of immediately ending the process. If this isn't feasible for some reason, maybe there should be an option to disable the quit menu item or at least trigger a prompt that warns users of potentially losing unsaved data before actually exiting.

I suppose you could think of this as a feature, but this is a common way to exit applications on MacOS and users will perceive it as a bug if the game is expected to have any kind of auto save functionality.

**Comment by DevinLeamy** [comment]
You're likely running into a race condition between when the `AppExit` event is created (e.g. `exit_on_all_closed`) and when you're consuming it. 

Unlike most other events, this order _actually matters_ because during the `Last` stage, if an `AppExit` event is received, then the subsequent frame will not run (the app will have closed), whereas typically during the next frame you'd have read the event that was previously missed.

The event is likely created in the `Last` stage by the `exit_on_all_closed` system. Therefore, to capture it the system reading the event should in the `Last` stage and after the `exit_on_all_closed` system.

From my testing (also on MacOS), the `.after(exit_on_all_closed)` doesn't seem to impact whether the event is captured by `on_exit_event`, but to be _certain_ that it will be captured it's probably a good idea to add it anyway.

```rust
fn main() {
    App::new()
        // ...
        .add_systems(Last, on_exit_event.after(exit_on_all_closed)) // or on_exit_event.after(exit_on_all_closed)
        .run();
}

fn on_exit_event(exit_events: EventReader<AppExit>) {
    if !exit_events.is_empty() {
        println!("Exit event received");
    }
}
```

As an alternative solution, we could move `exit_on_all_closed` into the `PostUpdate` stage, and then any system in the `Last` stage would certainly run after the `AppExit` event was written.

**Comment by mockersf** [comment]
> I suppose you could think of this as a feature, but this is a common way to exit applications on MacOS and users will perceive it as a bug if the game is expected to have any kind of auto save functionality.

As a side note, there are no guarantees and never will be that a "auto save on exit" feature will reliably work. It's not possible. Bevy will do its best to let you react to the event, Rust will do its best to call the drop implementations, but neither are guaranteed.

Auto save is usually implemented as "after significant events/actions" or "every x minutes".

**Comment by viridia** [comment]
I've run into this issue as well. Building an editor, I want to be able to prompt the user for unsaved changes; if Command-Q terminates the app with no recourse, this is a serious footgun, users are going to be very upset about losing hours of work because the app quit without giving them a chance to save.

**Comment by natepiano** [comment]
It may be difficult, but i still wish it were possible to reliably capture the application exiting - it's nonintuitive to me that clicking the close button allows me to capture AppExit but Command-Q does not.  I can understand not capturing panic's - but even then it seems we could - but a user initiated Command-Q? I just _expect_ it would be available.

My current use case is to capture window position and fullscreen mode so that wherever the user moves their window, i can save that on exit and then use it on the next launch. 

of course i can update this file any time that the user moves/resizes/fullscreen+/- but that seems incredibly inelegant as the only time it actually matters is when they quit the app.

**Comment by drewbluewasabi** [comment]
Is there a reason we can't have an event that fires on app exiting, that observers can hook into?
In my testing (on macOS and iOS), a simple `world.trigger(ExitingEvent)` put before cleanup works just fine as it's outside of the run loop.

I believe though, that a controlled quit as mentioned by @viridia cannot be solved as directly due an issue in Winit. From what I've seen, the default menu bar has 'Quit' set up to terminate. Even in the Winit examples, Cmd+Q sends the `Destroyed` event which cannot be recovered from.

**Comment by viridia** [comment]
Note that saving of preferences isn't the the only use case for intercepting app exit. A familiar example is the "quit without saving" confirmation dialog.

**Comment by drewbluewasabi** [comment]
> Note that saving of preferences isn't the the only use case for intercepting app exit. A familiar example is the "quit without saving" confirmation dialog.

Yeah, so a last-second event on app exiting as I mentioned would definitely be too much a hack.
A change is needed in Winit, to adjust the behavior to a window event. Then Bevy could expose that event to be handled. I’m not too sure how intentional Winit’s design for this is.
From what I’ve seen, this behavior was kindof thrown in with other changes for basic menu bar support.

There have also been issues brought up in other projects caused by this behavior: https://github.com/slint-ui/slint/issues/4482

**Comment by ibrahimsbashir** [comment]
I have just come across this issue too from trying to implement a "quit without saving" dialog and to cleanly terminate a UDP connection on Command + Q

I'll look into winit code to see if I can try and fix it and submit a PR, but wondering if anyone had workarounds for this in the meantime?

**Comment by drewbluewasabi** [comment]
> but wondering if anyone had workarounds for this in the meantime?

My raw solution doesn't seem to directly solve your problem, but here it is regardless:

Currently, I've just made a basic fork of Bevy, inserting an Observer Event fire directly before bevy terminates.
My hack doesn't allow you to cancel the termination, nor does it allow the Bevy event loop to continue (effectively freezes the user-facing app).

Here's a copy-paste of my workaround
On Bevy 0.18.1:

/crates/bevy_winit/src/lib.rs:
```rust
/// FROM MANUAL PATCH
///
/// Event fires during app exit. App exit yields for this event.
#[derive(Debug, Event)]
pub struct AppExitingEvent;
```

/crates/bevy_winit/src/state.rs: 
```rust
// \/ \/ Inside the function here \/ \/
impl ApplicationHandler<WinitUserEvent> for WinitAppRunnerState {
    // ...
    fn exiting(&mut self, _event_loop: &ActiveEventLoop) {
        // Patch code:
        println!("PATCH: Fired exiting event");
        self.world_mut().trigger(crate::AppExitingEvent);

        // ...
```
